File: GO/Sudoku.go
package main

import ("fmt";"strconv";"time")


/*------------------------CONST------------------------------*/

const S int = 9
const SS int = S*S
const R int = 3
const ALL Value = (1 << uint(S)) - 1

type Value uint


/* This is the main struct of the program.
 * It stores the board, a list with all possible values for each
 * row / col / sqr and a list of the remaning indices in the board.
 * It stores each value as a power of 2, the posibles values are just
 * |s of all the possibilities
 */
type Sudoku struct{
	board [SS]Value
	colsAv [S]Value
	rowsAv [S]Value
	sqrsAv [S]Value
}

/*-------------------------GLOBAL----------------------------*/

//To count the nodes and changes made
var forcedChanges int
var calls int

var remeaning [SS]int
var counter int

/*---------------------INITIALIZATION------------------------*/


func GenSud() (s Sudoku){
	/* Generates the sudoku, it is hard coded */
	temp_counter := 0
	
	sud := [S][S]Value{
		{0, 2, 4,  0, 0, 0,  0, 0, 0},
		{0, 0, 0,  0, 0, 7,  1, 0, 0},
		{0, 9, 0,  0, 0, 0,  0, 0, 0},
		{0, 0, 0,  0, 0, 0,  0, 8, 4},
		{0, 0, 0,  0, 7, 5,  0, 0, 0},
		{6, 0, 0,  0, 3, 0,  0, 0, 0},
		{0, 0, 0,  4, 0, 0,  0, 2, 9},
		{0, 0, 0,  2, 0, 0,  3, 0, 0},
		{1, 0, 0,  0, 0, 0,  0, 0, 0}} //17

	for i, l := range sud{
		for j, e := range l{
			if e == 0{
				s.board[index(i, j)] = 0
				remeaning[temp_counter] = index(i, j)
				temp_counter++
			}else{
				s.board[index(i, j)] = 1 << uint(e - 1)
			}
		}
	}
	counter = temp_counter

	return s
}

/*---------------------HELPER------------------------*/

func log2(v Value) int{
	/* Returns 1 + log2(n) or 0*/
	if v == 0{
		return 0
	}else{
		var log int
		for v != 1{
			log++
			v >>=1
		}
		return 1 + log
	}
}



func index(i, j int) int{
	return S*i + j
}
func coord(i int) (int, int){
	return i / S, i % S
}
func getSqrIndex(i, j int) int{
	return R*(i/R) + (j/R)
}
func isPow2(v Value) bool{
	return v & (v-1) == 0 && v != 0
}


/*---------------------CHECKING------------------------*/


func (self Sudoku) PrintSudoku() string{
	/* String representation of the Sudoku. TODO: Implement it as a to_string() */
	str := "--------------------\n"
	for i := 0; i < S; i++ {
		sSmall := ""
		for j := 0; j < S; j++ {

			if self.board[index(i, j)] == 0{
				sSmall += "- "
			}else{
				sSmall += strconv.Itoa(log2(self.board[index(i, j)])) + " "
			}

			if j % R == R - 1{
				sSmall += " "
			}
		}

		str += sSmall
		str += "\n"
		
		if i % R == R - 1 && i != S - 1{
			str += "\n"
		}
	}

	return str + "--------------------" + strconv.Itoa(self.filledTiles()) + "/" + strconv.Itoa(SS)
}

func (self Sudoku) filledTiles() int{
	/* Returns the number of filled tiles */
	t := 0
	for i := 0; i < SS; i++ {
		if self.board[i] != 0{
			t++
		}
	}
	return t
}

func (self Sudoku) finished() bool{
	/* Returns true if all the tiles are set, false otherwise */
	for i := 0; i < SS; i++ {
		if self.board[i] == 0{
			return false
		}
	}

	return true
}

func (self Sudoku) verifySudoku() bool{
	/* Returns if a sudoku is valid */
	var sum Value
	if ! self.finished(){
		return false
	}
	//Check rows
	for i := 0; i < S; i++ {
		sum = 0
		for j := 0; j < S; j++ {
			if sum & self.board[index(i, j)] != 0{
				return false
			}
			sum += self.board[index(i, j)]
		}
		if sum != ALL{
			return false
		}
	}
	//Check cols
	for i := 0; i < S; i++ {
		sum = 0
		for j := 0; j < S; j++ {
			if sum & self.board[index(j, i)] != 0{
				return false
			}
			sum += self.board[index(j, i)]
		}
		if sum != ALL{
			return false
		}
	}
	//Check sqrs
	for x := 0; x < R; x++ {
		for y := 0; y < R; y++ {
			sum = 0
			for i := 0; i < R; i++ {
				for j := 0; j < R; j++ {
					if sum & self.board[index(R*x + i, R*y + j)] != 0{
						return false
					}
					sum += self.board[index(R*x + i, R*y + j)]		
				}
			}
			if sum != ALL{
				return false
			}
		}
	}

	return true
}

/*--------------------------SOLVER-----------------------------*/

func (self Sudoku) canBeFinished() bool{
	/* Returns if a sudoku can be finished. true DOES NOT imply it can be finished.
	 * a sudoku cant be finished if one of the remaning tiles to fill has got no
	 * possibilities
	 */
	for i := 0; i < counter; i++{
		if self.board[remeaning[i]] == 0 && self.possible(remeaning[i]) == 0{
			return false
		}
	}

	return true
}


func (self Sudoku) possible(index int) Value{
	/* Returns the possible values for a given index */
	i, j := coord(index)
	return self.rowsAv[i] & self.colsAv[j] & self.sqrsAv[getSqrIndex(i, j)]
}

func setPossible(s *Sudoku){
	/* Initializes rows, cols, sqrs with the possibilities */
	var tempRows, tempCols, tempSqrs [S]Value
	var val Value

	for i := 0; i < S; i++ {
		for j := 0; j < S; j++ {
			val = s.board[index(i, j)]
			tempRows[i] |= val
			tempCols[j] |= val
			tempSqrs[getSqrIndex(i, j)] |= val
		}
	}

	for i := 0; i < S; i++ {
		s.rowsAv[i] = ALL ^ tempRows[i]
		s.colsAv[i] = ALL ^ tempCols[i]
		s.sqrsAv[i] = ALL ^ tempSqrs[i]
	}
}

func updateOne(index int, s *Sudoku){
	/* Removes board[index] from the possible values in its row / col / sqr */
	i, j := coord(index)

	s.rowsAv[i] &= ALL ^ s.board[index]
	s.colsAv[j] &= ALL ^ s.board[index]
	s.sqrsAv[getSqrIndex(i, j)] &= ALL ^ s.board[index]
}


func setForced(s *Sudoku) int{
	/* Does 1 iteration through the board setting all the tiles with just 1 possibility
	 * Returns:
	 *	0 if there were no changes
	 *	1 if there was at least one change
	 *	2 if the board is unsolvable
	 */
	var i, j int
	var val int
	var available, mask Value
	var updated int

	for k := 0; k < counter; k++{
		val = remeaning[k]
		if s.board[val] != 0{
			continue
		}
		i, j = coord(val)
		available = s.rowsAv[i] & s.colsAv[j] & s.sqrsAv[getSqrIndex(i, j)]

		if available == 0{
			return 2
		}

		if isPow2(available){
			s.board[val] = available
			mask = ALL ^ available
			
			s.rowsAv[i] &= mask
			s.colsAv[j] &= mask
			s.sqrsAv[getSqrIndex(i, j)] &= mask

			updated = 1
			
			forcedChanges++ //DEBUG

		}
	}

	return updated
}

func setAllForced(s *Sudoku) bool{
	/* Calls set_forced until there are no more forced tiles.
	 * Returns false if the board is unsolvable
	 */
	for lastUpdate := 1; lastUpdate == 1;{
		lastUpdate = setForced(s)
		if lastUpdate == 2{
			return false
		}
	}
	return true
}


func (self Sudoku) solve() (bool, Sudoku){
	/*
	* Main algorithm, tries to set all the forced values
	* When that is no longer possible, it branches and checks if
	* the resulting sudokus are solvable, it tries to reach as deep as
	* possible in each branch. It is recursive. Returns false, _ if the
	* sudoku is invalid
	*/
	calls++

	isPos := setAllForced(&self)

	if ! isPos{
		return false, self
	}
	if self.finished(){
		return true, self
	}
	
	//Faster in some cases
	var index int
	for i := counter - 1; self.board[index] != 0 && i >= 0; i-- {
		index = remeaning[i]	
	}
	
	//Alternative, some further testing is required to determine which one to use
	/*
	for i := 0; s.board[index] != 0 && i < counter; i++ {
		index = remeaning[i]	
	}
	*/
	var allPos Value = self.possible(index)
	var val Value
	for i := 0; i < S; i++ {
		val = (1 << uint(i))
		if val & allPos == 0{
			continue
		}

		newS := Sudoku(self) //Duplicates s
		newS.board[index] = val
		updateOne(index, &newS)
		
		if newS.canBeFinished(){ //This reduces the number of branches by about 1/4
			pos, newS := newS.solve()
			if pos{
				return true, newS
			}
		}
	}

	return false, Sudoku{}
}

func SolveMaster(s Sudoku) (Sudoku, /*ok*/ bool, /*forcedChanges*/ int, /*calls*/ int){
	rs := Sudoku(s)

	setPossible(&rs)

	if ! rs.canBeFinished(){
		return s, false, 0, 0
	}
	_, rs = rs.solve()

	ok := rs.filledTiles() == SS && rs.verifySudoku()

	return rs, ok, forcedChanges, calls
}

func main(){

	s := GenSud()
	fmt.Println(s.PrintSudoku())
	
	start := time.Now()
	
	s, ok, forcedChanges, calls := SolveMaster(s)

	end := time.Now()

	fmt.Println(s.PrintSudoku())

	fmt.Println("\nTotal changes:", forcedChanges)
	fmt.Println("Total nodes:", calls)
	fmt.Println("Is valid:", ok)
	fmt.Println("Time:", end.Sub(start))
	
}

File: Python/Sudoku.py
R = 3
S = 9
SS = S*S
ALL = (1 << S) - 1

class Sudoku(object):
    """This is the main class of the program.
    It stores the board, a list with all possible values for each
    row / col / sqr and a list of the remaning indices in the board.
    It stores each value as a power of 2, the posibles values are just
    |s of all the possibilities"""
    def __init__(self, _board, _rows, _cols, _sqrs, remeaning=None):
        self.board = _board
        self.rows = _rows
        self.cols = _cols
        self.sqrs = _sqrs

        if remeaning:
            self.remeaning = remeaning
        else:
            self.remeaning = [i for i in range(SS) if _board[i] == 0]

    def __repr__(self):
        """
        String representation of the sudoku
        """
        s = "--------------------\n"
        for i in range(S):
            s_small = ""
            for j in range(S):
                if self.board[index(i, j)] == 0:
                    s_small += "- "
                else:
                    s_small += str(log2(self.board[index(i, j)])) + " "

                if j % R == R - 1:
                    s_small += " "
            s += s_small + "\n"
            if i % R == R - 1 and i != S - 1:
                s += "\n"

        return s

    def duplicate(self):
        return Sudoku(list(self.board), list(self.rows), list(self.cols), list(self.sqrs), list(self.remeaning))

    def possible(self, index):
        """
        Returns the possible values for a tile
        """
        i, j = coord(index)
        return self.rows[i] & self.cols[j] & self.sqrs[get_sqr_index(i, j)]

    def update_one(self, index):
        """
        Removes the value in board[index] from the possible in its row / col / sqr
        """
        i, j = coord(index)
        self.rows[i] &= ALL ^ self.board[index]
        self.cols[j] &= ALL ^ self.board[index]
        self.sqrs[get_sqr_index(i, j)] &= ALL ^ self.board[index]

    def set_forced(self):
        """
        Sets all the tiles which have 1 possible value in one iteration through the board
        Returns:
           0 if there were no changes
           1 if at least one tile was updated
           2 if the board is unsolvable        
        """
        next_rem = []
        updated = 0

        for index in self.remeaning:
            i, j = coord(index)
            available = self.rows[i] & self.cols[j] & self.sqrs[get_sqr_index(i, j)]
            
            if available == 0:
                return 2

            if is_pow_2(available):
                self.board[index] = available
                self.rows[i] &= ALL ^ available
                self.cols[j] &= ALL ^ available
                self.sqrs[get_sqr_index(i, j)] &= ALL ^ available

                updated = 1
            else:
                next_rem.append(index)


        self.remeaning = list(next_rem)

        return updated

    def set_all_forced(self):
        """
        Calls set_forced until it stops making changes or the board is unsolvable
        """
        last_update = 1
        while last_update == 1:
            last_update = self.set_forced()
            if last_update == 2:
                return False
        return True

    def is_finished(self):
        """
        Returns True if all the tiles are set, False otherwise
        """
        for i in self.board:
            if i == 0:
                return False
        return True

def generate_sudoku():
    """
    Generates the sudoku, so far it is hard coded
    """
    board = list(map(get_pow, 
        [0, 2, 4,  0, 0, 0,  0, 0, 0, 
        0, 0, 0,  0, 0, 7,  1, 0, 0,
        0, 9, 0,  0, 0, 0,  0, 0, 0,
        0, 0, 0,  0, 0, 0,  0, 8, 4,
        0, 0, 0,  0, 7, 5,  0, 0, 0,
        6, 0, 0,  0, 3, 0,  0, 0, 0,
        0, 0, 0,  4, 0, 0,  0, 2, 9,
        0, 0, 0,  2, 0, 0,  3, 0, 0,
        1, 0, 0,  0, 0, 0,  0, 0, 0])) #17

    rows = [ALL] * S
    cols = [ALL] * S
    sqrs = [ALL] * S


    for i in range(S):
        for j in range(S):
            val = board[index(i, j)]
            rows[i] ^= val
            cols[j] ^= val
            sqrs[get_sqr_index(i, j)] ^= val

    return Sudoku(board, rows, cols, sqrs)

def log2(n):
    """
    Returns 1 + log2(n) or 0
    """
    if n == 0:
        return 0
    else:
        log = 0
        while n != 1:
            log += 1
            n >>= 1

        return 1 + log

def get_pow(n):
    return 0 if n == 0 else 1 << (n - 1)

def index(i, j):
    return S * i + j

def coord(i):
    return i // S, i % S

def get_sqr_index(i, j):
    return R*(i // R) + (j // R)

def is_pow_2(v):
    return (v & (v - 1)) == 0 and (v != 0)

def ls_of_possible(v):
    """
    Parses the possible values into a list of powers of 2
    """
    return [1<<i for i in range(S) if (v & (1 << i)) != 0]

def solve(sud):
    """
    Main algorithm, tries to set all the forced values
    When that is no longer possible, it branches and checks if
    the resulting sudokus are solvable, it tries to reach as deep as
    possible in each branch 
    """
    is_pos = sud.set_all_forced()

    if not is_pos:
        return False, None
    elif sud.is_finished():
        return True, sud

    index = sud.remeaning.pop()

    for value in ls_of_possible(sud.possible(index)):
        new_sud = sud.duplicate()
        new_sud.board[index] = value
        new_sud.update_one(index)

        pos, new_sud = solve(new_sud)
        if pos:
            return True, new_sud

    return False, None

def main():
    sudoku = generate_sudoku()
    print(sudoku)
    _, res = solve(sudoku)
    print(res)


if __name__ == '__main__':
    main()